### PR TITLE
Add `DataFrame.from_dict` classmethod

### LINKED ISSUE #9009: Make it easier to manually create DataFrames
We often need to create [Minimal Complete Verifiable Examples](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) with Dask DataFrames for blog posts, questions, etc.

Here's how Dask allows you to make these examples currently:

```python
import pandas as pd

pandas_df = pd.DataFrame(
    {"num1": [1, 2, 3, 4], "num2": [7, 8, 9, 10]},
)
df = dd.from_pandas(pandas_df, npartitions=2)
```

This is confusing for some users.  They are not sure why pandas needs to be imported and can struggle making the pandas => Dask connection.

The following syntax would make the MCVEs shorter and easier to understand:

```python
import dask.dataframe as dd

ddf = dd.DataFrame(
    {"num1": [1, 2, 3, 4], "num2": [7, 8, 9, 10]},
    npartitions=2,
)
```

Let me know if this suggestion sounds good and I can submit a PR.  Or perhaps there is already a way to directly create a Dask DataFrame in a similar manner?

### PR DESCRIPTION
**Pull request by MrPowers**
- [X] Closes #9009 
- [X] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

I'm not sure this was added in the right files / properly, so feel free to provide feedback!  Thank you!

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by quasiben** [comment]
add to allowlist

File: dask/dataframe/io/io.py
Comment (douglasdavis) [inline dask/dataframe/io/io.py:157]: Adding some line breaks is probably best
```suggestion
        The "orientation" of the data. If the keys of the passed dict
        should be the columns of the resulting DataFrame, pass 'columns'
        (default). Otherwise if the keys should be rows, pass 'index'.
        If 'tight', assume a dict with keys
        ['index', 'columns', 'data', 'index_names', 'column_names'].
```

File: dask/dataframe/io/io.py
Comment (douglasdavis) [inline dask/dataframe/io/io.py:161]:
```suggestion
        Column labels to use when ``orient='index'``. Raises a ValueError
        if used with ``orient='columns'`` or ``orient='tight'``.
```

File: dask/dataframe/io/io.py
Comment (douglasdavis) [inline dask/dataframe/io/io.py:152]: `npartitions` shouldn't be optional here (since you're using `from_pandas` and one must provide `npartitions` to `from_pandas` if `chunksize` is not used). Another route would be to support `chunksize` as an argument.

**Review by douglasdavis — CHANGES_REQUESTED** [review]
This indeed looks like a nice convenience (and covers more area of the pandas API). I had a few comments.

**Comment by pavithraes** [comment]
@MrPowers Thanks for this PR, it looks great! I think we can also add this function to the docs here:
- https://github.com/dask/dask/blob/main/docs/source/dataframe-api.rst#create-dataframes, and
- https://github.com/dask/dask/blob/main/docs/source/dataframe-create.rst#api

File: dask/dataframe/io/io.py
Comment (MrPowers) [reply dask/dataframe/io/io.py:152]: @douglasdavis - thanks for reviewing, good catch.  I made `npartitions` a required argument.  I think all users of `dd.from_dict` should manually set the partitions.  Thanks again!

**Comment by MrPowers** [comment]
@pavithraes - good catch about adding this method to the docs.  I think I added it properly.  Can you take a look and confirm?  Thanks!

File: docs/source/dataframe-create.rst
Comment (pavithraes) [inline docs/source/dataframe-create.rst:36]: Just a thought, not a blocker -- I think this might be better suited under a new heading "Python Dictionary" or "Python Collections", but `from_array` also doesn't fit perfectly under "file formats", so maybe we're fine.

File: dask/dataframe/io/io.py
Comment (pavithraes) [reply dask/dataframe/io/io.py:152]: @MrPowers Thanks for updating this! We may also need to remove "optional" from the doctstring:
```suggestion
    npartitions : int
```

File: docs/source/dataframe-create.rst
Comment (pavithraes) [inline docs/source/dataframe-create.rst:36]: Just a thought, maybe `from_dict` might be better suited under a different heading "Python Collections"?

This works as-is too because `from_array` also doesn't perfectly fit under "file formats". I'll leave it to you. :)

Comment (MrPowers) [reply docs/source/dataframe-create.rst:36]: @pavithraes - this is a good point.  I'm actually not entirely sure how `from_array` works, so not sure the best categorization.  I'm guessing it's not referring to a Dask Array cause there is a separate `from_dask_array` method.  I guess I'd argue to keep this as-is, mainly cause I don't know enough to have an opinion.

Comment (pavithraes) [reply docs/source/dataframe-create.rst:36]: Sounds good!

**Comment by pavithraes** [comment]
@MrPowers Looking at the CI failure, I think there might be some merge conflicts -- would you mind merging `main` into this branch and checking for any conflicts?

**Comment by MrPowers** [comment]
@pavithraes - I rebased on top of the latest version of main, let's see if that does the trick ;)

File: dask/dataframe/io/io.py
Comment (jrbourbeau) [inline dask/dataframe/io/io.py:149]: I see we're adding this as a top-level `dd.from_dict` method, but the corresponding pandas method is actually a classmethod -- so `pd.DataFrame.from_dict` instead of `pd.from_dict`. My default is usually to match the existing DataFrame API if one exists. Thoughts of making this `dd.DataFrame.from_dict`?

**Comment by jrbourbeau** [comment]
> The [failing windows-py3.10 test](https://github.com/dask/dask/runs/6306465028?check_suite_focus=true#step:6:21420) seems unrelated to this PR

Yep, I agree this looks unrelated. Just opened up https://github.com/dask/dask/issues/9035

**Comment by MrPowers** [comment]
Updated `from_dict` to be a class method based on a suggestion from @jrbourbeau.  `dd.DataFrame.from_dict` is more consistent with `pd.DataFrame.from_dict`.

@pavithraes - can you please take another look and check to make sure that I properly updated the docs to reflect that this is a class method 🙏 

Do you think this method should have the `@derived_from(pd.DataFrame)` annotation?  Thanks for all the help!!

**Comment by pavithraes** [comment]
> @pavithraes - can you please take another look and check to make sure that I properly updated the docs to reflect that this is a class method 🙏

@MrPowers Thank you! The docs updates look good to me.

 > Do you think this method should have the @derived_from(pd.DataFrame) annotation?

I don't think we need it here because you've added a nice and comprehensive docstring. :)

**Review by jrbourbeau — COMMENTED** [review]
Thanks for the updates @MrPowers! Didn't get a chance to look at them today. Will plan to review / merge tomorrow though

**Review by jrbourbeau — APPROVED** [review]
Thanks @MrPowers -- will merge after CI passes

**Pull request merged**
